workflow/util/util.go

```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
+		// special case where the retry node is the container of a containerSet
+		mustFind = wfv1.NodeTypePod
+	}
+
+	findBoundaries := false
+	for {
+
+		if curr == nil {
+			break
+		}
+
+		switch curr.n.Type {
+		case wfv1.NodeTypePod:
+			//ignore
+		case wfv1.NodeTypeContainer:
+			//ignore
+		case wfv1.NodeTypeSteps:
+			addToReset(curr.n.ID, false)
+			findBoundaries = true
+		case wfv1.NodeTypeStepGroup:
+			addToReset(curr.n.ID, false)
+			mustFind = wfv1.NodeTypeSteps
+		case wfv1.NodeTypeDAG:
+			addToReset(curr.n.ID, false)
+			findBoundaries = true
+		case wfv1.NodeTypeTaskGroup:
+			addToReset(curr.n.ID, false)
+			mustFind = wfv1.NodeTypeDAG
+		case wfv1.NodeTypeRetry:
+			addToReset(curr.n.ID, false)
+		case wfv1.NodeTypeSkipped:
+			// ignore -> doesn't make sense to reach this
+		case wfv1.NodeTypeSuspend:
+			// ignore
+		case wfv1.NodeTypeHTTP:
+			// ignore
+		case wfv1.NodeTypePlugin:
+			addToReset(curr.n.ID, false)
 		}
-		if !containsNode(resetParentGroupNodes, currentNodeID) {
-			newWF.Status.Nodes.Set(currentNodeID, resetNode(*currentNode.DeepCopy()))
-			resetParentGroupNodes = append(resetParentGroupNodes, currentNodeID)
-			log.Debugf("Reset connected group node %s", currentNode.Name)
+
+		if mustFind == "" && !findBoundaries {
+			curr = curr.parent
+			continue
 		}
-		if currentNode.BoundaryID != "" && currentNode.BoundaryID != oldWF.ObjectMeta.Name {
-			parentNode, err := oldWF.Status.Nodes.Get(currentNode.BoundaryID)
+
+		if findBoundaries {
+			curr, err = consumeBoundaries(curr, addToReset)
 			if err != nil {
-				log.Panicf("unable to obtain node for %s, panicking", currentNode.BoundaryID)
-			}
-			if isGroupNode(*parentNode) {
-				currentNodeID = parentNode.ID
-			} else {
-				break
+				return nil, nil, err
 			}
-		} else {
+			findBoundaries = false
+			continue
+		}
+
+		switch mustFind {
+		case wfv1.NodeTypePod:
+			curr, err = consumePod(curr, addToReset, addToDelete)
+		case wfv1.NodeTypeSteps:
+			curr, err = consumeSteps(curr, addToReset)
+		case wfv1.NodeTypeStepGroup:
+			curr, err = consumeStepGroup(curr, addToReset)
+		case wfv1.NodeTypeDAG:
+			curr, err = consumeDAG(curr, addToReset)
+		case wfv1.NodeTypeTaskGroup:
+			curr, err = consumeTaskGroup(curr, addToReset)
+		default:
+			return nil, nil, fmt.Errorf("invalid mustFind of %s supplied", mustFind)
+		}
+		mustFind = ""
+		if err != nil {
+			return nil, nil, err
+		}
+
+	}
+	return nodesToReset, nodesToDelete, nil
+}
+
+func setUnion[T comparable](m1 map[T]bool, m2 map[T]bool) map[T]bool {
+	res := make(map[T]bool)
+
+	for k, v := range m1 {
+		res[k] = v
+	}
+
+	for k, v := range m2 {
+		if _, ok := m1[k]; !ok {
+			res[k] = v
+		}
+	}
+	return res
+}
+func shouldRetryFailedType(nodeTyp wfv1.NodeType) bool {
+	if nodeTyp == wfv1.NodeTypePod || nodeTyp == wfv1.NodeTypeContainer {
+		return true
+	}
+	return false
+}
+
+// dagSortedNodes sorts the nodes based on the order they were created, omits onExitNode
```

Comment:
Technically not in creation order as children are all appended batched without regard to order.

If `task-1` finishes later than `task-2` then task-1's children may  get created later than task-2's but that won't be reflected here.

Reply:
Yeah its more, topological sorted order. Will change that.